server/cloudapps/expand.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package cloudapps
+
+import (
+	pluginapi "github.com/mattermost/mattermost-plugin-api"
+	"github.com/mattermost/mattermost-plugin-cloudapps/server/configurator"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+type ExpandEntity string
+
+const (
+	ExpandActingUser = ExpandEntity("ActingUser")
+	ExpandUser       = ExpandEntity("User")
+	ExpandChannel    = ExpandEntity("Channel")
+	ExpandConfig     = ExpandEntity("Config")
+)
+
+type ExpandLevel string
+
+const (
+	ExpandAll     = ExpandLevel("All")
+	ExpandSummary = ExpandLevel("Summary")
+)
+
+type Expand struct {
+	ActingUser ExpandLevel
+	Channel    ExpandLevel
+	Config     bool
+	User       ExpandLevel
+}
+
+type Expanded struct {
+	ActingUser *model.User
+	Channel    *model.Channel
+	Config     *MattermostConfig
+	User       *model.User
+}
+
+type MattermostConfig struct {
+	SiteURL string
+}
+
+type Expander struct {
+	mm           *pluginapi.Client
+	configurator configurator.Configurator
+
+	ActingUser *model.User
+	Channel    *model.Channel
+	Config     *model.Config
+	User       *model.User
+}
+
+func NewExpander(mm *pluginapi.Client, configurator configurator.Configurator) *Expander {
+	return &Expander{
+		mm:           mm,
+		configurator: configurator,
+	}
+}
+
+func (e *Expander) Expand(expand *Expand, actingUserID, userID, channelID string) (expanded *Expanded, err error) {
```

Comment:
Should expansions be scoped? i.e. Do we want to limit what certain apps can expand and how much can they expand it?

Reply:
Added a note to the doc, 1/5 out of scope for this project. But the exact consent model is TBD. E.g. admin consents that the app has access to "UserJoinedChannel", but maybe the channel admin's consent is required to Subscribe to any specific channel?